### PR TITLE
Fix line ending bug

### LINKED ISSUE #89: Literal and folded strings cannot handle "\n \n" at the end of a string
Adding the test case `'whitespace between line breaks at the end\n \n'` to `test/string_test.dart` causes tests to fail.

```diff
diff --git a/test/string_test.dart b/test/string_test.dart
index b92c20a..a18f57a 100644
--- a/test/string_test.dart
+++ b/test/string_test.dart
@@ -10,6 +10,7 @@ final _testStrings = [
   "this is a fairly' long string with\nline breaks",
   'whitespace\n after line breaks',
   'whitespace\n \nbetween line breaks',
+  'whitespace between line breaks at the end\n \n',
   '\n line break at the start',
   'word',
   'foo bar',
```

Tests failures:
```console
$ dart test
00:01 +400 -1: loading test/golden_test.dart                                                                                                                                                                                           
Successfully tested 20 inputs against golden files, created 0 golden files
00:01 +400 -1: test/string_test.dart: Root FOLDED string (4) [E]                                                                                                                                                                       
  Assertion failed: (package:yaml_edit) Modification did not result in expected result.
  
  # YAML before edit:
  > 
  
  # YAML after edit:
  > >-
  >   whitespace between line breaks at the end
  >    
  >   
  
  Please file an issue at:
  https://github.com/dart-lang/yaml_edit/issues/new?labels=bug
  
  package:yaml_edit/src/editor.dart 579:7   YamlEditor._performEdit
  package:yaml_edit/src/editor.dart 249:14  YamlEditor.update
  test/string_test.dart 39:20               main.<fn>
  

To run this test again: dart test test/string_test.dart -p vm --plain-name 'Root FOLDED string (4)'
00:01 +415 -2: test/string_test.dart: Root LITERAL string (4) [E]                                                                                                                                                                      
  Assertion failed: (package:yaml_edit) Modification did not result in expected result.
  
  # YAML before edit:
  > 
  
  # YAML after edit:
  > |-
  >   whitespace between line breaks at the end
  >    
  >   
  
  Please file an issue at:
  https://github.com/dart-lang/yaml_edit/issues/new?labels=bug
  
  package:yaml_edit/src/editor.dart 579:7   YamlEditor._performEdit
  package:yaml_edit/src/editor.dart 249:14  YamlEditor.update
  test/string_test.dart 39:20               main.<fn>
  

To run this test again: dart test test/string_test.dart -p vm --plain-name 'Root LITERAL string (4)'
00:10 +624 -2: Some tests failed.                                                                                                                                                                                                      

Consider enabling the flag chain-stack-traces to receive more detailed exceptions.
For example, 'dart test --chain-stack-traces'.
```

### PR DESCRIPTION
**Pull request by kekavc24**
I debugged #89 further while submitting #91 and noted several issues that are closely coupled:

- Trailing line-break only applied to `YamlList` or `YamlMap` and never a `YamlScalar`
- Dangling comments and line-breaks left behind if a comment spans multiple lines.

This PR:
1. Applies a line break after each `YamlNode` encoded within a block.
2. Fixes #89.
3. Introduces a `normalizeEncodeBlock` function that dynamically prunes the additional line break if:
    - The deepest `YamlScalar` isn't encoded as `ScalarStyle.LITERAL` or `ScalarStyle.FOLDED`. Also fixes #89
    - The old node didn't have a trailing line-break.
4. Introduces a `skipAndExtractComments` function which greedily skips any comments and whitespace belonging to the `YamlNode` being replaced.

Further changes made in #93

---

- [x] I’ve reviewed the contributor guide and applied the relevant portions to this PR.

<details>
  <summary>Contribution guidelines:</summary><br>

- See our [contributor guide](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md) for general expectations for PRs.
- Larger or significant changes should be discussed in an issue before creating a PR.
- Contributions to our repos should follow the [Dart style guide](https://dart.dev/guides/language/effective-dart) and use `dart format`.
- Most changes should add an entry to the changelog and may need to [rev the pubspec package version](https://github.com/dart-lang/sdk/blob/main/docs/External-Package-Maintenance.md#making-a-change).
- Changes to packages require [corresponding tests](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md#Testing).

Note that many Dart repos have a weekly cadence for reviewing PRs - please allow for some latency before initial review feedback.
</details>

### CODE REVIEW

**Comment by kekavc24** [comment]
@jonasfj I have improved this PoC further by:

1. Lazily looking ahead for comments as that would bring the draft closer to the how `YamlEditor` currently does its edits.
2. Refactored the functions that edit `YamlNode` that are encoded as block elements to take advantage of the `skipAndExtractComments` which makes code a bit straight-forward and easy to understand (entirely subjective until you take a look) 😅 

All in all, the existing and currently failing tests kinda vouch for this PoC but I'm ready to push its limit to prove its worth with stronger and more convoluted tests.

**Comment by jonasfj** [comment]
I'm down for taking a look at this.
Currently, traveling and I fear this requires a bit of attention to understand. So I'll try to take a look next week.

I saw you had nice comments explaining some of the methods. That's awesome.

Feel free to give some review hints in the PR description :)

Also if any of these changes could be done independently, please do consider submitting smaller independent PRs. It's much easier to review, which often leads to faster turnaround time.

**Comment by kekavc24** [comment]
I've tried to split the 22 commits over 3 PRs based on the changes I made.  `#90 -> #93 -> #94`.

File: lib/src/equality.dart [hunk 1]
```diff
@@ -87,8 +87,10 @@ int deepHashCode(Object? value) {
 }
 
 /// Returns the [YamlNode] corresponding to the provided [key].
```
Comment (jonasfj) [inline lib/src/equality.dart:89]: nit: consider updating the documentation.

Also feel free to submit this as a separate PR, and hint that you plan on using it in this PR. Then we can reduce the size of this PR a bit.

File: lib/src/utils.dart [hunk 2]
```diff
@@ -273,6 +273,196 @@ String getLineEnding(String yaml) {
   return windowsNewlines > unixNewlines ? '\r\n' : '\n';
 }
 
+/// Extracts comments for a node that is replaced within a [YamlMap] or
+/// [YamlList] or a top-level [YamlScalar] of the [yaml] string provided.
+///
+/// [currentEndOffset] represents the end offset of [YamlScalar] or [YamlList]
+/// or [YamlMap] being replaced, that is, `end + 1`.
```
Comment (jonasfj) [inline lib/src/utils.dart:280]: This is probably better illustrated with an example.

```dart
/// ```dart
/// final yaml = 'my_key: "replaced-value"\n';
/// //                                    ^--- currentEndOffset points to the newline
/// fimal currentEndOffset = 24;
/// ```
```
ofcourse, I might have counted wrong here :rofl: 

Also does this mean that `currentEndOffset` can be greater than the length of the `yaml` document? (`currentEndOffset == yaml.length`, so `yaml[currentEndOffset]` might throw)

**Example**: `k: v` will have `currentEndOffset: 4` (if I'm replacing `v`, right).

Comment (jonasfj) [reply lib/src/utils.dart:280]: Sorry, if I'm asking dumb questions here, but it might be worth making some examples.

File: lib/src/utils.dart [hunk 3]
```diff
@@ -273,6 +273,196 @@ String getLineEnding(String yaml) {
   return windowsNewlines > unixNewlines ? '\r\n' : '\n';
 }
 
+/// Extracts comments for a node that is replaced within a [YamlMap] or
+/// [YamlList] or a top-level [YamlScalar] of the [yaml] string provided.
```
Comment (jonasfj) [inline lib/src/utils.dart:277]: Isn't it just always a for a `YamlNode` that is replaced?

Is it only if replaced, or if removed too? (is there a difference? An example might help).

If this can be: "a node within a `YamlMap` or `YamlList`, or a top-level `YamlScalar`".
Then isn't that the same as saying: `YamlNode`?
Can a YamlNode appear elsewhere in the document?

Just trying to be concise and understand if we listed out the options this way for a reason, or if we could write the equivalent thing more concisely.

Oh, I guess the node that `currentEndOffset` points to can't be a top-level `YamlList` or top-level `YamlMap` is that correct? (or am I reading it too formal).
If so, maybe it'd be shorter to write that. Or write both.

File: lib/src/utils.dart [hunk 4]
```diff
@@ -273,6 +273,196 @@ String getLineEnding(String yaml) {
...
+///
+/// [currentEndOffset] represents the end offset of [YamlScalar] or [YamlList]
+/// or [YamlMap] being replaced, that is, `end + 1`.
+///
+/// [nextStartOffset] represents the start offset of the next [YamlNode].
+/// May be null if the current [YamlNode] being replaced is the last node
+/// in a [YamlScalar] or [YamlList] or if its the only top-level [YamlScalar].
+/// If not sure of the next [YamlNode]'s [nextStartOffset] pass in null and
+/// allow this function to handle that manually.
```
Comment (jonasfj) [inline lib/src/utils.dart:286]: I think it's better we know why it's appropriate to use this, rather than: I wasn't sure.

If not passing it always works, then why would we ever pass it in (I don't think we should be overly concerned with performance here, if that's the concern).

File: lib/src/utils.dart [hunk 5]
```diff
@@ -273,6 +273,196 @@ String getLineEnding(String yaml) {
...
 
+/// Extracts comments for a node that is replaced within a [YamlMap] or
+/// [YamlList] or a top-level [YamlScalar] of the [yaml] string provided.
+///
+/// [currentEndOffset] represents the end offset of [YamlScalar] or [YamlList]
+/// or [YamlMap] being replaced, that is, `end + 1`.
+///
+/// [nextStartOffset] represents the start offset of the next [YamlNode].
+/// May be null if the current [YamlNode] being replaced is the last node
+/// in a [YamlScalar] or [YamlList] or if its the only top-level [YamlScalar].
```
Comment (jonasfj) [inline lib/src/utils.dart:284]: Just curious, but does this make assumptions about block-mode?

File: lib/src/editor.dart [hunk 6]
```diff
@@ -241,11 +241,13 @@ class YamlEditor {
 
     if (path.isEmpty) {
       final start = _contents.span.start.offset;
-      final end = getContentSensitiveEnd(_contents);
+      var end = getContentSensitiveEnd(_contents);
       final lineEnding = getLineEnding(_yaml);
-      final edit = SourceEdit(
-          start, end - start, yamlEncodeBlock(valueNode, 0, lineEnding));
-
+      end = skipAndExtractCommentsInBlock(_yaml, end, null, lineEnding).$1;
```
Comment (jonasfj) [inline lib/src/editor.dart:246]:
```suggestion
      final (endOffset, _) = skipAndExtractCommentsInBlock(_yaml, end, null, lineEnding);
      end = endOffset;
```
or similar, but let's avoid `.$1` if we can.


Or make `skipAndExtractCommentsInBlock` return a record with named properties. So we can do: `skipAndExtractCommentsInBlock(...).endOffset;`.

File: lib/src/utils.dart [hunk 7]
```diff
@@ -273,6 +273,196 @@ String getLineEnding(String yaml) {
...
+
+  if (terminalNode is! YamlScalar) {
+    loop:
+    while (terminalNode is! YamlScalar) {
+      switch (terminalNode) {
+        case YamlList list:
+          {
+            if (list.isEmpty) {
+              terminalNode = list;
+              break loop;
```
Comment (jonasfj) [inline lib/src/utils.dart:419]: Just return

File: lib/src/utils.dart [hunk 8]
```diff
@@ -273,6 +273,196 @@ String getLineEnding(String yaml) {
...
+String normalizeEncodedBlock(
+  String yaml,
+  String lineEnding,
+  int nodeToReplaceEndOffset,
+  YamlNode update,
+  String updateAsString,
+) {
+  var terminalNode = update;
+
+  if (terminalNode is! YamlScalar) {
```
Comment (jonasfj) [inline lib/src/utils.dart:411]: Maybe make a utility method called `findTerminalNode(YamlNode node)`

Then we can properly document what a terminal node is too.

Also if you do that, you won't need to `break loop;` you can just do a `return` instead.

File: lib/src/utils.dart [hunk 9]
```diff
@@ -273,6 +273,196 @@ String getLineEnding(String yaml) {
...
+    loop:
+    while (terminalNode is! YamlScalar) {
+      switch (terminalNode) {
```
Comment (jonasfj) [inline lib/src/utils.dart:414]: Consider something like `while(true) { switch(terminalNode) {` instead.

Also, how does this work? Don't the switch-statement have to be exhaustive?
I get that it can only be `YamlList` and `YamlMap`, when it's not `YamlScalar`.
But the compiler can't see that because `YamlNode` isn't a `sealed` class.

Maybe add a `default` case that throws an `AssertionError`. It should never happen, but if someone makes a new kind of `YamlNode` in the future, it's better than whatever else might happen.

File: lib/src/utils.dart [hunk 10]
```diff
@@ -273,6 +273,196 @@ String getLineEnding(String yaml) {
...
+/// `YamlScalar` and removes any unwanted line-break within the
+/// [updateAsString].
+///
+/// This is achieved by obtaining the chunk of the [yaml] that is after the
+/// current node being replaced using its [nodeToReplaceEndOffset]. If:
+///   1. The chunk has any trailing line-break then the it is left untouched.
+///   2. The node being replaced with [update] is not the last node, then it
+///      is left untouched.
+///   3. The terminal node in [update] is a `YamlScalar`, that is,
+///      the last [YamlNode] within the [update] that is not a collection.
```
Comment (jonasfj) [inline lib/src/utils.dart:401]: Please consider adding some example.

Also why is it that we don't do this normalization when we create the `updateAsString`? Wouldn't that be more correct -- I don't know -- I probably don't understand all the context here, that's why I'm asking :D

Comment (kekavc24) [reply lib/src/utils.dart:401]: That's the point of adding the dangling line-break. 

`yamlEncodeBlock` has no context itself of which method called it or the current structure of the yaml document. It could be:

1. `updateInList` or `_appendToBlockList` or `_insertInBlockList` for lists
2. `_replaceInBlockMap` or `_addToBlockMap` for block maps
3. `Yaml.update` for top level `YamlNode`.

Thus, the callers should make a call to the function to prune the trailing line break because they have an idea of the yaml's structure.

Additionally, I found it quite advantageous on our part to try and include line breaks for an existing `YamlNode` being replaced to edits. This reduces bugs and removes the need to try and guess where the next line break is. 

It better to include it and check later if it was there 🤷🏽‍♂️

File: lib/src/utils.dart (hunk 5 above)
Comment (kekavc24) [reply lib/src/utils.dart:284]: I wrote that function to be open ended and versatile in case someone wanted to tackle #71 or do something cool with `YamlEditor` if they have the next node in a `YamlList` or `YamlMap`

File: lib/src/utils.dart (hunk 4 above)
Comment (kekavc24) [reply lib/src/utils.dart:286]: I wrote that function to be open ended and versatile in case someone wanted to tackle #71 or do something cool with `YamlEditor` if they have the next node in a `YamlList` or `YamlMap`

File: lib/src/utils.dart (hunk 5 above)
Comment (kekavc24) [reply lib/src/utils.dart:284]: That was a heads-up for anyone who may decide to call that function. It should be `YamlMap` or `YamlList`.

File: lib/src/utils.dart (hunk 8 above)
Comment (kekavc24) [reply lib/src/utils.dart:411]: Sure. 

I'm beginning to think I shouldn't have split this PR to #93 and #94? All those changes address the title of this PR or simply build up to it.

File: lib/src/utils.dart (hunk 3 above)
Comment (kekavc24) [reply lib/src/utils.dart:277]: Oh, yeah. I'll change that. I wrote that method for #90 as a PoC before working on #94.

**Comment by kekavc24** [comment]
@jonasfj I am beginning to think I shouldn't have split this PR as #93 and #94 have some changes I made that minimally improve this PR.

Any change I make here will be a conflict for #93 and #94 that could have been avoided. I'll make a separate PR after #94 for any reviews you suggest or I could close #90 and #93 in favour of #94.

File: lib/src/utils.dart (hunk 2 above)
Comment (kekavc24) [reply lib/src/utils.dart:280]: > Also does this mean that `currentEndOffset` can be greater than the length of the `yaml` document? (`currentEndOffset == yaml.length`, so `yaml[currentEndOffset]` might throw)

`currentEndOffset` shouldn't be greater than `yaml.length` when passed in as an argument. May be I should throw if it is?

---
> 
> **Example**: `k: v` will have `currentEndOffset: 4` (if I'm replacing `v`, right).

I think `3` since it's always zero-based.

---

> Sorry, if I'm asking dumb questions here, but it might be worth making some examples.
> 
> And I think we should make sure we're doing it right, if we get he wrong invariants we'll have a lot of bugs to fix :D

Sure, will add examples.

**Comment by jonasfj** [comment]
Feel free to mark comments as "Resolved" once you've:
 * addressed the comment, or,
 * replied explaining why you don't want to do anything (not all comments need be addressed).

I think it would be useful to add some examples in the code. Especially for `skipAndExtractCommentsInBlock` and `normalizeEncodeBlock`.
I'm not sure I fully understand what they are even supposed to do. To be fair, maybe I just didn't look hard enough at it yet :rofl:

**Comment by kekavc24** [comment]
I've added commits I had moved to the other PRs. Without the changes, the `skipAndExtractCommentsInBlock` and `normalizeEncodedBlock` won't make sense. (see also failing tests for bonus points)

The existing code was only scanning for the first instance of `#` and `\n` but `yaml` allow for comments spanning multiple lines. A key issue was with blocks lists in yaml since it scanned for `-`. But comments can also have `-`? Case in point:

```yaml
# Existing code works great here for edits
- valueToRemove # Comment with hyphen (-)
- nextValue


# Existing code fails horribly here for edits

- valueToRemove # Comment with hyphen (-)
                     # but spanning (oops, hyphen "-")
                     #
                     # multiple lines (another hyphen "-")
- nextValue
```

`skipAndExtractCommentsInBlock`:

1. Scans ahead until it skips all comments whether it's a single line or spans multiple lines
2. Can do it `lazily` (remembers the first `\n` it saw while scanning ahead if no comments are present) but also `greedily` (skips both `\n` and/or comments)

I realized we can do it `lazily` when making updates within maps or lists i.e for `inserts` or `splices` etc. And `greedily` exclusively for removing elements in maps/lists.

`normalizeEncodedBlock` just helps us to avoid always falling back to `double quotes` even for string that can be styled as `folded` or `literal`

File: lib/src/utils.dart [hunk 11]
```diff
@@ -273,6 +273,272 @@ String getLineEnding(String yaml) {
...
+/// Do note that this function has no context of the structure of the [yaml]
+/// but assumes the caller does and requires comments based on the offsets
+/// provided and thus, may be erroneus since it exclusively scans for `#`
+/// delimiter or extracts the comments between the [endOfNodeOffset] and
+/// [nextStartOffset] if both are provided.
+///
+/// Returns the `endOffset` of the last comment extracted that is `end + 1`
+/// and a `List<String> comments`. It is recommended (but not necessary) that
+/// the caller checks the `endOffset` is still within the bounds of the [yaml].
+({int endOffset, List<String> comments}) skipAndExtractCommentsInBlock(
```
Comment (jonasfj) [inline lib/src/utils.dart:303]: Could you give me a few examples of how this works.

Or maybe a few unit tests, even if we don't keep them, I'd like to understand a bit more what the intention here is.

It seems like it takes a string slice, and returns:
 * Comments from within the slice (not sure why)
 * Offset where the next non-comment thing in the slice begins, or ends? (or did I misunderstand that?)

Comment (kekavc24) [reply lib/src/utils.dart:303]: > It seems like it takes a string slice, and returns:
> 
> * Comments from within the slice (not sure why)
> * Offset where the next non-comment thing in the slice begins, or ends? (or did I misunderstand that?)

Yes. It returns where the non-comment starts but with a nuanced approach based on the `bool` argument you pass into the `greedy` parameter.

----

> Could you give me a few examples of how this works.

Yeah, sure. See the examples below.


```yaml
- valueToRemove # Comment with hyphen (-)
                # but spanning (oops, hyphen "-")

- nextValue
```

- `greedy` as `true` returns the `offset` of the `-` indicating the start of the next value in the list.
- `greedy` as `false` returns the `offset` of the last `\n` (line-break) + 1 which is just after the last closing bracket `)`

---

> Or maybe a few unit tests, even if we don't keep them, I'd like to understand a bit more what the intention here is.

A nice example even without the unit tests would be an element in a list or map with multi-line comments being removed. For example:

Performing `YamlEditor(yaml).remove([0]);` on the string below:


```yaml
- valueToRemove # Comment with hyphen (-)
                     # but spanning (oops, hyphen "-")
                     #
                     # multiple lines (another hyphen "-")
- nextValue


# Expected output without this PR

                     # but spanning (oops, hyphen "-")
                     #
                     # multiple lines (another hyphen "-")
- nextValue


# Expected output with PR
- nextValue
```

The offending code can be found here which is similar to `_removeFromBlockMap`:
https://github.com/dart-lang/yaml_edit/blob/35f4248c7bbba289b3899fa55486e2f31ef1a8c5/lib/src/list_mutations.dart#L343-L374

Comment (jonasfj) [reply lib/src/utils.dart:303]: **Disclaimer:** These thoughts are not fully conceived, I'm still trying to understand.

----

So if the input for `YamlEditor(yaml).remove([0])` is:
```yaml
- valueToRemove      # Comment with hyphen (-)
                     # but spanning (oops, hyphen "-")
                     #
                     # multiple lines (another hyphen "-")
  # nextValue is awesome:
- nextValue
```
Then with this PR we'd get:
```yaml
- nextValue
```

Right?

----

I'm just highlighting this example, because detecting that a comment is spanning multiple lines is difficult.

It's kind of obvious to us humans that the comment `# nextValue is awesome:` is associated with `- nextValue` and that all the comments with equal indentation are related to `- valueToRemove`.

But if we tried to make a heuristic to detect this, such heuristic would fail if comments aren't perfectly aligned :see_no_evil: I can see no end of corner cases :rofl: 

----

So maybe we have to ask, when faced with `YamlEditor(yaml).remove([0])` on:
```yaml
# A
  # B
- valueToRemove # C
                # D
# E

# F
  # G
- nextValue # H
# I
```
What comments do we wish to remove / preserve?

At the top of my head, I'm thinking:
 * `H` and `I` should definition be preserved.
 * Removing `C` seems reasonable.
 * Do we want to treat `A` and `B` the same way?
   * Would we want to detect that `A` and `B` are different based on indentation?
 * Do we want to treat `D`, `E`, `F` and `G` the same way?
   * Would we want to detect that `D` and `E` are not the same based on indentation?
   * Would we want to detect that `E` and `F` are not the same based on empty line?
   * Would we want to detect that `F` and `G` are not the same based on indentation?

I'm suspecting that maybe it's simplest to say that:
 * `A` and `B` are treated the same way: we preserve them.
 * `D`, `E`, `F` and `G` are treated the same way: we preserve them.

It could be that it's best to say retain comments, unless they are inside the value we're removing (or on the same line). We risk leaving some half gabled comments, but which is worse:
 * Removing too many comments?
 * Retaining half a comment related to a value that's been removed?

If we make heuristics for all this to be more accurate, don't we risk things becoming extremely complicated? :rofl:

Comment (kekavc24) [reply lib/src/utils.dart:303]: > Right?

Yes.

Funnily enough, I totally understand your point of view 😄. That’s how my head was spinning before I wrote `skipAndExtractComments`.

However, I reduced this PR’s scope to actually do what `YamlEditor` currently WANTS to do while ensuring we can have `folded`/`literal` strings. This is still a PoC.

From the PR, you can see I didn’t change how `YamlEditor` works just nudged it to do what existing code wants to do.

Personally, I don’t see it as complexity but rather as a challenge to live up to the package’s description the **RIGHT WAY** since if we provide a way to preserve comments, wouldn’t it be great to at least provide a subjective way to remove them?

If you are open to it, I can come up with an issue by the end of the week with my view and where we can both float ideas on how to tackle this and once we settle on something I will submit a PR. If not, then please indicate what we can keep from this PR and what not to keep.

Comment (kekavc24) [reply lib/src/utils.dart:303]: > I'll have to admit that I'm not entirely sure what this PR does 🙈

1. We can now encode strings with a trailing line-break as `ScalarStyle.FOLDED` or `ScalarStyle.LITERAL` instead of falling back to double quotes.  See:
    - #89
    - #91
    
2. Because we can do `1` above, it comes with a cost. The edit we suggest may have a line-break already since we ensure we preserve any dangling line-breaks after calling `normalizeEncodedBlock` which:
    - Preserves line-breaks for `ScalarStyle.FOLDED` or `ScalarStyle.LITERAL` or if the value of the `YamlScalar` itself has a line-break
    - Checks if the node being replaced had a line-break and preserves it in the new encoded block
   
 The cost is to ensure we include any existing line-break in the edit without trying to introduce any breaking behavior to the current `YamlEditor` functionality. 
 
From my previous comment,  `YamlEditor` **CURRENTLY** wants to get rid of inline comments in the terminal node but does it poorly causing issues with edits with some of the tests actually enabling the bug. (See some of the failing tests when this PR is run. I’ll highlight them on request)

Thus, why I added `skipAndExtractComments` which skips the inline comments if that’s what `YamlEditor` wanted to do. 

I must admit, maybe the function is a bit overzealous and returns the comments it skipped but I thought why not.

Also, this PR was just a PoC. Nothing more than that.

Comment (jonasfj) [reply lib/src/utils.dart:303]: @kekavc24 how about you start an issue outline what we want to do.
And we try to reach consensus on that first.

I think perhaps it's best to try write tests and make small changes (many small PRs are infinitely better, than large ones). It's more work to split logic into separate PRs. But the challenge is also getting there without introducing bugs (actually, it's getting there while convincing maintainers of this package that we're not introducing new bugs).

**Comment by mosuem** [comment]
Closing as the [dart-lang/yaml_edit](https://github.com/dart-lang/yaml_edit) repository is merged into the [dart-lang/tools](https://github.com/dart-lang/tools) monorepo. Please re-open this PR there!
